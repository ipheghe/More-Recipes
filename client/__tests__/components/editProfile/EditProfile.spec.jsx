import expect from 'expect';
import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import { Provider } from 'react-redux';
import render from 'react-test-renderer';
import { HashRouter as Router } from 'react-router-dom';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import ConnectedEditProfile, { PureEditProfile }
  from '../../../src/components/EditProfile';
import mockItems from '../../__mocks__/mockItems';
import mockAuthCheck from '../../__mocks__/mockAuthCheck';


const middleware = [thunk];
const mockStore = configureMockStore(middleware);
const initialState = {
  auth: {
    error: 'Invalid username',
    message: '',
    userData: mockItems.user,
    authenticated: false,
    categories: []
  },
  user: {
    status: '',
    error: '',
    message: '',
    userData: mockItems.user,
  },
  category: {
    error: '',
    message: '',
    categoryList: [mockItems.category],
    userCategoryList: [mockItems.category]
  },
};

const store = mockStore(initialState);

const props = {
  errorMessage: '',
  updateUserRecord: jest.fn(() => Promise.resolve()),
  fetchUsername: jest.fn(() => Promise.resolve()),
  userData: mockItems.user
};

const state = {
  username: 'okon',
  fullName: 'abcde',
  mobileNumber: '234702388888',
  email: 'okon@yahoo.com',
  hasErrored: false,
  errorMessage: '',
  isLoading: false
};

const event = {
  preventDefault: jest.fn(),
  target: {
    username: 'okon',
    password: 'abcde',
    fullName: 'Okon Imo',
    mobile: '234702388888',
    email: 'okon@yahoo.com'
  }
};

jest.useFakeTimers();

/**
 *@description  setup function to mount component
 *
 * @return { * } null
 */
const setup = () => {
  const shallowWrapper = shallow(<PureEditProfile {...props} />);
  return {
    shallowWrapper
  };
};

describe('<PureEditProfile', () => {
  beforeEach(() => {
    mockAuthCheck();
  });

  it(`should render a loader component before 
  MyRecipes component receives needed props`, () => {
      props.userData = null;
      const shallowWrapper = shallow(<PureEditProfile {...props} />);
      shallowWrapper.setState({ isLoading: true });
      expect(shallowWrapper).toBeDefined();
      expect(shallowWrapper.find('Loader').length).toBe(1);
      expect(shallowWrapper.exists()).toBe(true);
    });

  it('renders without crashing', () => {
    const { shallowWrapper } = setup();
    shallowWrapper.setState(state);
    expect(shallowWrapper).toBeDefined();
    expect(shallowWrapper.find('EditProfileForm').length).toBe(1);
    expect(shallowWrapper.exists()).toBe(true);
  });

  it('should match component snapshot', () => {
    const tree = render
      .create(<Provider store={store}>
        <Router><ConnectedEditProfile {...props} /></Router>
      </Provider>);
    expect(tree).toMatchSnapshot();
  });

  it('calls handleChange event', () => {
    sinon.spy(PureEditProfile.prototype, 'handleChange');
    const { shallowWrapper } = setup(false);
    shallowWrapper.instance().handleChange(event);
    expect(PureEditProfile.prototype.handleChange.calledOnce).toEqual(true);
  });

  it(`calls componentWillReceiveProps if 
    userData from props is available`, () => {
      sinon.spy(PureEditProfile.prototype, 'componentWillReceiveProps');
      const { shallowWrapper } = setup();
      shallowWrapper.instance().componentWillReceiveProps(props);
      expect(PureEditProfile.prototype.componentWillReceiveProps.calledOnce)
        .toEqual(true);
    });

  it('calls handleUpdate event after update button is clicked', () => {
    sinon.spy(PureEditProfile.prototype, 'handleUpdate');
    const { shallowWrapper } = setup();
    shallowWrapper.instance().handleUpdate(event);
    expect(PureEditProfile.prototype.handleUpdate.calledOnce).toEqual(true);
  });

  it('displays error message if user inputs a null username value', () => {
    sinon.spy(PureEditProfile.prototype, 'validateFormField');
    const { shallowWrapper } = setup(false);
    state.username = '';
    shallowWrapper.setState(state);
    shallowWrapper.instance().validateFormField();
    expect(shallowWrapper.state().hasErrored).toEqual(true);
    expect(shallowWrapper.state().errorMessage)
      .toEqual('Username must start with a letter and have no spaces.');
    jest.runAllTimers();
    expect(shallowWrapper.state().errorMessage).toEqual('');
  });

  it(`displays error message if user inputs a 
     fullName value with length less than 4`, () => {
      const { shallowWrapper } = setup(false);
      state.username = 'okon';
      state.fullName = 'oko';
      shallowWrapper.setState(state);
      shallowWrapper.instance().validateFormField();
      expect(shallowWrapper.state().hasErrored).toEqual(true);
      expect(shallowWrapper.state().errorMessage)
        .toEqual('fullname must contain more than 3 chareacters');
      jest.runAllTimers();
      expect(shallowWrapper.state().errorMessage).toEqual('');
    });

  it('displays error message if user inputs an invalid mobile number', () => {
    const { shallowWrapper } = setup(false);
    state.mobileNumber = '';
    state.fullName = 'okon Akem';
    shallowWrapper.setState(state);
    shallowWrapper.instance().validateFormField();
    expect(shallowWrapper.state().hasErrored).toEqual(true);
    expect(shallowWrapper.state().errorMessage)
      .toEqual('mobile number must contain only numbers');
    jest.runAllTimers();
    expect(shallowWrapper.state().errorMessage).toEqual('');
  });

  it('displays error message if user inputs an invalid email', () => {
    const { shallowWrapper } = setup(false);
    state.mobileNumber = '234702388888';
    state.email = 'okon@ya';
    shallowWrapper.setState(state);
    shallowWrapper.instance().validateFormField();
    expect(shallowWrapper.state().errorMessage)
      .toEqual('Invalid Email Address');
    jest.runAllTimers();
    expect(shallowWrapper.state().errorMessage).toEqual('');
  });

  it('dispatches updateUser action after validatiing fields', () => {
    const { shallowWrapper } = setup();
    state.email = 'okon@yahoo.com';
    shallowWrapper.setState(state);
    shallowWrapper.instance().validateFormField();
    expect(PureEditProfile.prototype.validateFormField.calledOnce)
      .toEqual(false);
    expect(shallowWrapper.state().username).toEqual('okon');
  });

  it('calls toggleModalState method', () => {
    sinon.spy(PureEditProfile.prototype, 'toggleModalState');
    const { shallowWrapper } = setup();
    shallowWrapper.setState(state);
    shallowWrapper.instance().toggleModalState();
    expect(PureEditProfile.prototype.toggleModalState.calledOnce)
      .toEqual(true);
  });

  it('calls toggleModalStateOff method', () => {
    sinon.spy(PureEditProfile.prototype, 'toggleModalStateOff');
    const { shallowWrapper } = setup();
    shallowWrapper.setState(state);
    shallowWrapper.instance().toggleModalStateOff();
    expect(PureEditProfile.prototype.toggleModalStateOff.calledOnce)
      .toEqual(true);
  });
});
