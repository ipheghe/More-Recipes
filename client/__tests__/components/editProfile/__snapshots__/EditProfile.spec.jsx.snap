// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PureEditProfile should match component snapshot 1`] = `
<div>
  <div
    className="banner-background"
  >
    <div
      className="profile-background"
    >
      <div
        className="container"
      >
        <nav
          className="navbar navbar-toggleable-md navbar-light bg-faded"
        >
          <button
            aria-controls="navbarTogglerDemo01"
            aria-expanded="false"
            aria-label="Toggle navigation"
            className="navbar-toggler"
            data-target="#navbarTogglerDemo01"
            data-toggle="collapse"
            type="button"
          >
            <span
              className="navbar-toggler-icon"
            />
          </button>
          <div
            className="collapse navbar-collapse"
            id="navbarTogglerDemo01"
          >
            <a
              className="navbar-brand invisible-button"
              href="#/dashboard/top-recipes"
              onClick={[Function]}
            >
              <h2>
                Profile Page
              </h2>
            </a>
            <ul
              className="navbar-nav mr-auto mt-2 mt-lg-0"
            >
              <li
                className="nav-item"
              >
                <button
                  className="nav-link invisible-button disabled"
                />
              </li>
            </ul>
            <form
              className="form-inline my-2 my-lg-0"
            >
              <input
                className="form-control mr-sm-2"
                formNoValidate={true}
                name="keyword"
                onChange={[Function]}
                placeholder="Search"
                required={true}
                type="text"
                value=""
              />
              <button
                className="btn btn-outline-success my-2 my-sm-0"
                onClick={[Function]}
              >
                Search
              </button>
            </form>
          </div>
        </nav>
        <br />
        <div
          className="row profile-landing"
        >
          <section
            className="col-md-3 profile-details"
          >
            <main>
              <div
                className="div-profile"
              >
                <h5>
                  @
                  linda
                </h5>
                <p>
                  <span>
                    <i
                      aria-hidden="true"
                      className="fa fa-circle"
                      style={
                        Object {
                          "color": "green",
                        }
                      }
                    />
                  </span>
                  Online
                </p>
                <br />
                <br />
                <button
                  className="invisible-button nav-editProfile"
                  onClick={[Function]}
                  style={
                    Object {
                      "color": "#0275d8",
                      "textAlign": "center",
                    }
                  }
                >
                  Edit Profile
                </button>
                <button
                  className="invisible-button nav-changePassword"
                  onClick={[Function]}
                  style={
                    Object {
                      "color": "#0275d8",
                      "textAlign": "center",
                    }
                  }
                >
                  Change Password
                </button>
              </div>
              <br />
              <div
                className="div-profile"
              >
                <h5>
                  Add Category
                </h5>
                <div
                  className="form-group profile-form"
                >
                  <input
                    className="form-control"
                    id="category-name"
                    name="categoryName"
                    onChange={[Function]}
                    placeholder="Enter Name"
                    type="text"
                    value=""
                  />
                </div>
                <div>
                  <button
                    className="btn btn-success"
                    onClick={[Function]}
                    type="button"
                  >
                    Add
                  </button>
                </div>
                <br />
                <div
                  className="profile-category-button"
                >
                  <button
                    className="btn btn-default btn-sm"
                    name="categoryModalName"
                    onClick={[Function]}
                    type="button"
                    value={1}
                  >
                    Local Dish
                  </button>
                </div>
              </div>
              <div />
            </main>
          </section>
          <section
            className="col-md-9 profile-tabs"
          >
            <div
              className="edit-profile-div-section"
            >
              <br />
              <h3>
                <b>
                  Edit Profile
                </b>
              </h3>
              <br />
              <div>
                <form
                  className="reg-form"
                >
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="enterUsername"
                    >
                      Username:
                    </label>
                    <div
                      className="input-group"
                    >
                      <span
                        className="input-group-addon"
                      >
                        <i
                          className="fa fa-user"
                        />
                      </span>
                      <input
                        className="form-control"
                        name="username"
                        onChange={[Function]}
                        placeholder="e.g john112"
                        required={true}
                        type="text"
                        value="linda"
                      />
                    </div>
                  </div>
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="enterFullName"
                    >
                      Full Name:
                    </label>
                    <div
                      className="input-group"
                    >
                      <span
                        className="input-group-addon"
                      >
                        <i
                          className="fa fa-user-o"
                        />
                      </span>
                      <input
                        className="form-control"
                        name="fullName"
                        onChange={[Function]}
                        placeholder="e.g John James"
                        required={true}
                        type="text"
                        value="Linda George"
                      />
                    </div>
                  </div>
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="enterMobile"
                    >
                      Mobile:
                    </label>
                    <div
                      className="input-group"
                    >
                      <span
                        className="input-group-addon"
                      >
                        <i
                          className="fa fa-mobile"
                        />
                      </span>
                      <input
                        className="form-control"
                        name="mobileNumber"
                        onChange={[Function]}
                        placeholder="e.g 08034562838"
                        required={true}
                        type="number"
                        value="2348034526172"
                      />
                    </div>
                  </div>
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="enterEmail"
                    >
                      Email address:
                    </label>
                    <div
                      className="input-group"
                    >
                      <span
                        className="input-group-addon"
                      >
                        <i
                          className="fa fa-envelope"
                        />
                      </span>
                      <input
                        className="form-control"
                        name="email"
                        onChange={[Function]}
                        placeholder="e.g john@yahoo.com"
                        required={true}
                        type="email"
                        value="linda@gmail.com"
                      />
                    </div>
                  </div>
                  <div
                    className="edit-profile-button"
                  >
                    <button
                      className="btn btn-success"
                      onClick={[Function]}
                    >
                      Update
                    </button>
                    <a
                      href="#dashboard/top-recipes"
                    >
                      <button
                        className="btn btn-danger"
                        type="button"
                      >
                        Cancel
                      </button>
                    </a>
                  </div>
                </form>
              </div>
              <br />
            </div>
          </section>
        </div>
      </div>
    </div>
  </div>
</div>
`;
